common/src/chainstate/utxo.rs

```diff
@@ -0,0 +1,462 @@
+        &mut self,
+        utxo_entries: HashMap<OutPointKey, UtxoEntry>,
+        block_hash: H256,
+    ) -> Result<(), Error> {
+        for (key, entry) in utxo_entries {
+            let parent_entry = self.utxos.get(&key);
+
+            // Ignore non-dirty entries (optimization).
+            if entry.is_dirty {
+                match parent_entry {
+                    None => {
+                        // The parent cache does not have an entry, while the child cache does.
+                        // We can ignore it if it's both spent and FRESH in the child
+                        if !(entry.is_fresh && entry.is_spent()) {
+                            // Create the utxo in the parent cache, move the data up
+                            // and mark it as dirty.
+                            let mut entry_copy = entry.clone();
+                            entry_copy.is_dirty = true;
+
+                            self.utxos.insert(key, entry_copy);
+                            // TODO: increase the memory usage
+                        }
+                    }
+                    // found entry in the parent cache
+                    Some(parent_entry) => {
+                        if entry.is_fresh() && !parent_entry.is_spent() {
+                            // The utxo was marked FRESH in the child cache, but the utxo
+                            // exists in the parent cache. If this ever happens, it means
+                            // the FRESH flag was misapplied and there is a logic error in
+                            // the calling code.
+                            return Err(Error::UtxoAlreadyExists);
+                        }
+
+                        if parent_entry.is_fresh && entry.is_spent() {
+                            // The grandparent cache does not have an entry, and the utxo
+                            // has been spent. We can just delete it from the parent cache.
+                            self.utxos.remove(&key);
+                        } else {
+                            // A normal modification.
+                            let mut entry_copy = entry.clone();
+                            entry_copy.is_dirty = true;
+                            entry_copy.is_fresh = parent_entry.is_fresh;
+                            self.utxos.insert(key, entry_copy);
+                            // TODO: update the memory usage
+
+                            // NOTE: It isn't safe to mark the utxo as FRESH in the parent
+                            // cache. If it already existed and was spent in the parent
+                            // cache then marking it FRESH would prevent that spentness
+                            // from being flushed to the grandparent.
+                        }
+                    }
+                }
+            }
+        }
+
+        self.current_block_hash = Some(block_hash);
+        Ok(())
+    }
+
+    fn derive_cache(&self) -> UtxosCache {
+        self.clone()
```

Comment:
Clone? Why? We should create a new instant linked to the parent.

Reply:
updated,